src/raid.cpp

```diff
@@ -0,0 +1,923 @@
+        if (!chunkmac.finished)
+        {
+            chunkmac = transfer->chunkmacs[chunkid];
+            cipher->ctr_crypt(chunkstart, chunksize, startpos, transfer->ctriv, chunkmac.mac, false, !chunkmac.finished && !chunkmac.offset);
+            if (endpos == ChunkedHash::chunkceil(chunkid, transfer->size))
+            {
+                LOG_debug << "Finished chunk: " << startpos << " - " << endpos << "   Size: " << chunksize;
+                chunkmac.finished = true;
+                chunkmac.offset = 0;
+            }
+            else
+            {
+                LOG_debug << "Decrypted partial chunk: " << startpos << " - " << endpos << "   Size: " << chunksize;
+                chunkmac.finished = false;
+                chunkmac.offset += chunksize;
+            }
+        }
+        chunkstart += chunksize;
+        startpos = endpos;
+        endpos = ChunkedHash::chunkceil(startpos, finalpos);
+        chunksize = static_cast<unsigned>(endpos - startpos);
+    }
+}
+
+
+bool RaidBufferManager::tryRaidHttpGetErrorRecovery(unsigned errorConnectionNum)
+{
+    if (isRaid())
+    {
+        raidHttpGetErrorCount[errorConnectionNum] += 1;
+
+        g_faultyServers.add(tempurls[errorConnectionNum]);
+
+        unsigned errorSum = 0;
+        for (unsigned i = RAIDPARTS; i--; )
+        {
+            errorSum += raidHttpGetErrorCount[i];
+        }
+
+        if (errorSum < 3)
+        {
+            if (unusedRaidConnection < RAIDPARTS)
+            {
+                LOG_warn << "5 connection cloudraid shutting down connection " << errorConnectionNum << " due to error, and starting " << unusedRaidConnection << " instead";
+
+                // start up the old unused connection, and cancel this one.  Other connections all have real data since we were already in 5 connection mode
+                clearOwningFilePieces(raidinputparts[unusedRaidConnection]);
+                clearOwningFilePieces(raidinputparts[errorConnectionNum]);
+                raidrequestpartpos[unusedRaidConnection] = raidpartspos;
+                raidrequestpartpos[errorConnectionNum] = raidpartspos;
+            }
+            else
+            {
+                LOG_warn << "6 connection cloudraid shutting down connection " << errorConnectionNum << " due to error";
+                clearOwningFilePieces(raidinputparts[errorConnectionNum]);
+                raidrequestpartpos[errorConnectionNum] = raidpartspos;
+            }
+
+            unusedRaidConnection = errorConnectionNum;
+            return true;
+        }
+    }
+    return false;
+}
+
+bool RaidBufferManager::connectionRaidPeersAreAllPaused(unsigned slowConnection)
+{
+    if (!isRaid())
+    {
+        return false;
+    }
+
+    // see if one connection is stalled or running much slower than the others, in which case try the other 5 instead
+    for (unsigned j = RAIDPARTS; j--; )
+    {
+        if (j != slowConnection && j != unusedRaidConnection && !connectionPaused[j])
```

Comment:
If the goal is to check whether the other *5* connections (excluding `slowConnection`) are paused in order to continue with those 5 instead of 6... I think it could happen that there's already one `unusedRaidConnection`, so already using 5 connections. If others than `slowConnection` are all paused, it may return `true` and, in consequence, would we start using 4 connections?
The case is as follows. Imagine you're checking `slowConnection = 0`, and the `unusedRaidConnection = 1`. Connections 2, 3, 4 and 5 are all paused. This code would return `true`.

Reply:
I reply myself: the `unusedRaidConnection` will be used again, and the `slowConnection` would be the new `unusedRaidConnection` :smiley:

Reply:
:-)